Task3/Task3_Jordan_Petrov/src/python/car_wash/dbpool.py

```diff
@@ -0,0 +1,159 @@
+"""DB pool manager.
+It is module that describes steps to connect to DB.
+It uses pool of connection to avoid load on DB side.
+"""
+
+import logging
+import time
+from contextlib import contextmanager
+from threading import Lock
+
+import psycopg2
+
+from src.python.car_wash.config import READ_ONLY_USER, READ_WRITE_USER
+from utils import SingletonMeta
+
+
+class DBPoolBase(metaclass=SingletonMeta):
+    """Base class of DB pool."""
+
+    def _create_connection(self) -> psycopg2.connect:
+        """Create connection to DB"""
+        raise NotImplementedError('No code added into _create_connection method')
+
+    def _get_connection(self) -> psycopg2.connect:
+        """Get connection from pool"""
+        raise NotImplementedError('No code added into _get_connection method')
+
+    def _put_connection(self, conn: psycopg2.connect) -> None:
+        """Put connection back to pool"""
+        raise NotImplementedError('No code added into _put_connection method')
+
+
+class RODBPool(DBPoolBase):
+    """Read only pool of connections."""
+
+    def __init__(self, hostname: str, password: str, username: str, _dbname: str, pool_size: int) -> None:
+        """Constructor of RODBPool connections
+
+        :Parameters:
+            - `hostname` - DB hostname
+        """
+        super().__init__()
+        self._pool = []
+        self._hostname = hostname
+        self._password = password
+        self._username = username
+        self._dbname = _dbname
+        self._pool_size = pool_size
+        self._lock = Lock()
+
+    def _create_connection(self) -> psycopg2.connect:
+        """Create connection to DB"""
+        return psycopg2.connect(dbname=self._dbname,
+                                user=self._username,
+                                password=self._password,
+                                host=self._hostname)
+
+    def _get_connection(self) -> psycopg2.connect:
+        """Get connection from pool"""
+
+        while True:
+            with self._lock:
+                if len(self._pool) > 0:
+                    return self._pool.pop()
+                else:
+                    time.sleep(1)
+
+    def _put_connection(self, conn: psycopg2.connect) -> None:
+        """Put connection back to pool"""
+        with self._lock:
+            self._pool.append(conn)
+
+    @contextmanager
+    def connection(self):
+        """
+        Context manager for the read only pool.
+        Use cursor to execute queries.
+        """
+        if len(self._pool) < self._pool_size:
```

Comment:
The idea is correct. Please check my comment: https://github.com/mehalyna/CarWashing-System/pull/38/files#r928253372